src/python/pants/backend/python/providers/pyenv/rules.py

```diff
@@ -0,0 +1,183 @@
+        relative path in the downloaded archive, e.g. `./bin/protoc`.
+        """
+        return f"./pyenv-{self.version}/bin/pyenv"
+
+
+class PyenvPythonProvider(PythonProvider):
+    pass
+
+
+PYENV_NAMED_CACHE = ".pyenv"
+PYENV_APPEND_ONLY_CACHES = FrozenDict({"pyenv": PYENV_NAMED_CACHE})
+
+
+@rule
+async def get_python(
+    request: PyenvPythonProvider,
+    python_setup: PythonSetup,
+    platform: Platform,
+    pyenv_subsystem: PyenvSubsystem,
+) -> PythonExecutable:
+    env_vars, pyenv = await MultiGet(
+        Get(EnvironmentVars, EnvironmentVarsRequest(["PATH", "LDFLAGS"])),
+        Get(DownloadedExternalTool, ExternalToolRequest, pyenv_subsystem.get_request(platform)),
+    )
+
+    python_to_use = request.interpreter_constraints.minimum_python_version(
+        python_setup.interpreter_versions_universe
+    )
+    if python_to_use is None:
+        raise ValueError(
+            f"Couldn't determine a compatible Interpreter Constraint from {python_setup.interpreter_versions_universe}"
+        )
+
+    which_python_result = await Get(
+        ProcessResult,
+        Process(
+            [pyenv.exe, "latest", "--known", python_to_use],
+            input_digest=pyenv.digest,
+            description=f"Choose specific version for Python {python_to_use}",
+            env={"PATH": env_vars.get("PATH", "")},
+            # Caching the result is OK, since if the user really needs a different patch,
+            # they should list a more precise IC.
+        ),
+    )
+    specific_python = which_python_result.stdout.decode().strip()
+
+    shim_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "install_python_shim.sh",
+                    dedent(
+                        f"""\
+                        #!/usr/bin/env bash
+                        set -e
+                        export PYENV_ROOT=$(realpath {PYENV_NAMED_CACHE})
+                        DEST="$PYENV_ROOT"/versions/{specific_python}
+                        DONE="$DEST"/DONE
+                        LOCK="$DEST"/DONE.lock
+                        # Use double-check locking around the pyenv install
+                        if [ ! -f "$DONE" ]; then
+                            mkdir -p "$DEST" 2>/dev/null || true
+                            exec {{lock_fd}}>"$LOCK"
+                            flock -x "$lock_fd"
+                            if [ ! -f "$DONE" ]; then
+                                {pyenv.exe} install {specific_python}
+                                # Removing write perms helps ensure users aren't accidentally modifying Python
+                                # or the site-packages
+                                chmod -R -w "$DEST"
+                                chmod +w "$DEST"
+                                touch "$DONE"
+                            fi
+                        fi
+                        echo $(realpath "$DEST"/bin/python)
```

Comment:
I said I'd stay out, but no one seems to have noticed this.

So, you have a problem:
```
$ ssh gha@pantsbuild-gha-mac-M1 realpath --version
zsh:1: command not found: realpath
$ ssh gha@pantsbuild-gha-mac-M1 flock --version
zsh:1: command not found: flock
$ ssh gha@pantsbuild-gha-mac-x86-64 realpath --version
zsh:1: command not found: realpath
$ ssh gha@pantsbuild-gha-mac-x86-64 flock --version
zsh:1: command not found: flock
```

This is not even to mention if the machine doesn't have the pyenv Python build requirements. The failures in these cases are likely pretty cryptic. It would be good to make sure that, from the perspective of someone running `./pants ...` who doesn't have one or more of these things, that the error is semi-grokkable.

Reply:
Hmm correct me if I'm wrong, but wouldn't the `#!/usr/bin/env bash` at the top mean your OS would run the script through bash? Those lines seem like it's `zsh`.

Agreed on the build deps, but not sure the best way to handle. (I wonder if the pyenv/Python error is already grokkable?)

Reply:
They are not a bash builtins either - they're binaries on your system or not independent of the shell. Re-running with bash -c '...' gives the same answer - Macs don't have these binaries by default.

Reply:
Okie dokie, looks like I'll be swtiching to a Python-based script then. Thanks for the callout! Bummer the tests aren't being run on Mac (it seems :thinking: )

Reply:
I think `@pytest.mark.platform_specific_behavior` is required for that.

Reply:
But don't let that fool you either. It means ~nothing. GHA Macs have alot of non-default software installed, for example. The only real solution when using shell to implement Pants goals is to research well.

Reply:
That seems odd though - you need python to install python? Then I can just ensure the right python up front and not need any of this.

Reply:
You should be able to use `sys.executable` to run a Python script that just uses `subprocess` and `fcntl` just fine. It's an abstraction leak to use the Pants host Python like this; so it should be called out in a comment.

Reply:
But you could also just implement realpath with readlink I think and go back to the polling solution.

Reply:
Thanks for the tip, but I've taken your Slack comment(s) to heart and just going to assume `sys.executable` is off limits.

> That seems odd though - you need python to install python? Then I can just ensure the right python up front and not need any of this.

We already have this restriction today. There must be _some_ kind of Python that exists so we can run non-user code. The two main instances being rule-code that needs to `gunzip` and `pex` itself (to select ICs). Look for the `find_python` rule.

Whether that code continues as is, becomes `sys.executable`, or something else is up to a future PR :smile:

Reply:
`sys.executable` should be off-limits for user code, but this is not that.

Reply:
Either way that's an implementation detail of `find_python`, and we could discuss that _not_ in this PR, as this PR will just use `PythonBinary` :wink:

Reply:
We can discuss in a separate issue, but note that PythonBinary uses the `python-bootstrap` subsystem to find python, and that is tangled up with interpreter discovery for user code. Specifically [here](https://github.com/benjyw/pants/blob/a8d9179426c2e46a1f368e04b869a66df486026f/src/python/pants/backend/python/util_rules/pex_environment.py#L174). 

Probably `PythonBinary` should just be hardwired to `sys.executable` and bypass all the rest of it.

Reply:
Yeah that's my point 😀

Reply:
Actually, no we can't. `sys.executable` is for the environment host :smile: